React section: Adds Class-based components lesson (#24805)
TheOdinProject/curriculum: react/class_components/class_based_components.md

### Introduction

All the components by far, have been functional in style and syntax. This is common now, but you will see a different `class` based syntax too. In this lesson, we explore how a class-based component is written and the significance of the different elements seen in one.

### Lesson Overview

This section contains a general overview of topics that you will learn in this lesson.

- Learn the structure of a class component and how they are written.
- How to use props and state in class components.
- Highlight the uses of `this` in class components.

> ### The What's And Why's

Author: Grammar gosh.. [Here's what I found](https://english.stackexchange.com/questions/146348/the-whys-and-the-hows), so I'm simply dropping them. Thanks for that, learnt something today-

1.  Implement a delete button for each task. The delete button should remove that specific task from the state array, thus deleting the task itself! Styling isn't a priority at this moment, but the button tag should be styled by default.

2.  Implement a new class component, `Count` that displays the count of the number of todos, at any given time.

3.  Implement an edit button for each task. It should replace the todo with an input field, and change the button itself to 'Resubmit', so the edits can be saved. This is a comparatively harder task, kudos for when you finish it!

<iframe src='https://codesandbox.io/embed/component-comparisions-difgyb?fontsize=14&hidenavigation=1&theme=dark'
        style='width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;'
        title='Component comparisions'
        allow='accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking'
        sandbox='allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts'

> ></iframe>

Reviewer: I've created a sandbox on TOP's account, we should change to that link instead: 
```
<iframe src="https://codesandbox.io/embed/github/TheOdinProject/react-examples/tree/main/class-component-example?fontsize=14&hidenavigation=1&theme=dark"
     style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;"
     title="react-example"
     allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
     sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
   ></iframe>
```

### Introduction

All the components by far, have been functional in style and syntax. This is common now, but you will see a different `class` based syntax too. In this lesson, we explore how a class-based component is written and the significance of the different elements seen in one.

### Lesson Overview

This section contains a general overview of topics that you will learn in this lesson.

> - A note on the relevance of class components.
> - Covers the structure of a class component, the uses of the constructor and render method.
> - Discusses the ways to use props and state in class components.
> - Highlights the uses of `this` in class components.

Reviewer: We can simplify the Lesson Overview:

```suggestion

- Learn the structure of a class component and how they're written.
- How to use props and state in class components.
- Highlight the uses of `this` in class components.
```

### The What's And Why's

> In your previous lessons, you have already been introduced to functional components, and the basic patterns in which components get written now a days. However, React components did not look this way since React was introduced.

Reviewer: ```suggestion
In your previous lessons, you have already been introduced to functional components, and the basic patterns in which components get written now a days. However, React components did not look this way when React was introduced.
```

If you look into any older React codebase, you'll notice a lot of classes. These are known as class-based components. Prior to February 2019, functional components were also called state-less, as there was no way to manage state in them. This was changed when hooks were introduced, leading to less verbose and 'neater' components.

> In your career, chances are, you will be dealing with legacy code, so there will be days where you would be dealing with class components. To meet this issue, we take a peek into the works and intricacies of a class-based component, and how they work.

Reviewer: Wording suggestion:

```suggestion
In your career, chances are, you will be dealing with legacy code, so there will be days where you would be dealing with class components. Let's peek into the intricacies of a class-based component, and how they work.
```

import React, { useState } from "react";

const FunctionalInput = ({ name }) => {
  /*
    We declare two state variables and their setters,
    one to store the To-Do's
    and the other to store the value of the input field
  */
  const [todos, setTodos] = useState(["Just some demo tasks", "As an example"]);
  const [inputVal, setInputVal] = useState("");

  const handleInputChange = (e) => {
    setInputVal(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    setTodos((todo) => [...todo, inputVal]);
    setInputVal("");
  };

  return (
    <section>
      <h3>{name}</h3>
      {/* The input field to enter To-Do's */}
      <form onSubmit={handleSubmit}>
        <label htmlFor="task-entry">Enter a task: </label>
        <input
          type="text"
          name="task-entry"
          value={inputVal}
          onChange={handleInputChange}
        />
        <button type="submit">Submit</button>
      </form>
      <h4>All the tasks!</h4>
      {/* The list of all the To-Do's, displayed */}
      <ul>
        {todos.map((todo) => (
          <li key={todo}>{todo}</li>
        ))}
      </ul>
    </section>
  );
};

export default FunctionalInput;

> ~~~

Reviewer: We should remove comments in this code example. By this point, learners should be well versed in functional components and how state works.

> That was a solid chunk of code. Take a while, sip some water and read it a couple of times.

Reviewer: <details>
<summary></summary>

![image](https://media4.giphy.com/media/sIDxMqTWbVxOM6gRPQ/giphy.gif?cid=ecf05e47u05ku7ojdqsilc04d7zsi957lv5kikj598sxswbh&rid=giphy.gif&ct=g)

          <input
            type="text"
            name="task-entry"
            value={this.state.inputVal}
            onChange={this.handleInputChange}
          />
          <button type="submit">Submit</button>
        </form>
        <h4>All the tasks!</h4>
        {/* The list of all the To-Do's, displayed */}
        <ul>
          {this.state.todos.map((todo) => (
            <li key={todo}>{todo}</li>
          ))}
        </ul>
      </section>
    );
  }
}

export default ClassInput;
~~~

And there we go, we have successfully made our first class-based component, as easy as that!

### Assignment

<div class="lesson-content__panel" markdown="1">

For the purposes of this assignment, we take the simple class-based component that we built. There is an interactive example provided at the end, where you can work on.

1.  Implement a delete button for each task. The delete button should remove that specific task from the state array, thus deleting the task itself! Styling isn't a priority at this moment, but the button tag should be styled by default.

2.  Implement a new class component, `Count` that displays the count of the number of todos, at any given time.

> 3.  Implement an edit button for each task. It should replace the todo with an input field, and change the button itself to 'Resubmit', so the edits can be saved. This is a comparatively harder task, kudos for when you finish it!

Reviewer: I like the assignment section. Hands on practice is invaluable!
